### PR TITLE
TypeScript.md update

### PR DESCRIPTION
**Pull request by PopGoesTheWza**
Fixes no specific issue.

Attempt to document the `export`/`import` limitation  while proposing possible workarounds.

- [ ] `npm run test` succeeds.
- [ ] `npm run lint` succeeds.
- [ ] Appropriate changes to README are included in PR.

Since it is solely a `.md` PR, above boxes are irrelevent.

### CODE REVIEW

File: docs/typescript.md [hunk 1]
```diff
@@ -21,12 +22,14 @@ This quickstart guide describes how to create a TypeScript project from scratch.
 
 ### Prerequisites
 
-1. Ensure you have upgrade to clasp >= 1.5.0
+1. Ensure you have upgraded to clasp version 1.5.0 or more
```
Comment (grant) [inline docs/typescript.md]:
```
1. Ensure you have upgraded to clasp's version to  >= 1.5.0.
```

File: docs/typescript.md [hunk 2]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
 
 ### TypeScript Support
 
-Currently, `clasp` supports [`typescript@2.9.2`](https://www.npmjs.com/package/typescript/v/2.9.2). If there is a feature in a newer  TypeScript version that you'd like to support, or some experimental flag you'd like enabled, please file a bug.
+Currently, `clasp` supports [`typescript@3.3.3`](https://www.npmjs.com/package/typescript/v/3.3.3). If there is a feature in a newer TypeScript version that you'd like to support, or some experimental flag you'd like enabled, please open an issue.
+
+### Modules, exports and imports
+
+Currently, Google Apps Script does not support modules, hence the typical `export`/`import` patern cannot be used and the example below will fail:
```
Comment (grant) [inline docs/typescript.md]: Apps Script does not support ES Modules. Hence, the typical `export`/`import` pattern cannot be used and the example below will fail:

File: docs/typescript.md [hunk 3]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
+import { foo } from './module';  // this statement is ignored
+
+const bar = foo;  // the variable `foo` does not exist in the global namespace
+```
+
+There are some possible workaround though:
+
+#### The `exports` declaration workaround
+
+The idea here is to make TypeScript aware of the `exports` object and its imported content.
```
Comment (grant) [inline docs/typescript.md]: This workaround makes TypeScript aware of the `exports` object and its imported content.

File: docs/typescript.md [hunk 4]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
+declare namespace exports {
+  const foo: _i_foo;  // declare typed imported content to the proper namespcae
+}
+
+exports.foo;  // address important content as it will be visible when transpiled
+```
+
+#### The `namespace` statement workaround
+
+Here we take advantage of TypeScript "namespaces" (formerly known as "internal module") to achieve code isolation.
```
Comment (grant) [inline docs/typescript.md]: This workaround takes advantage of TypeScript "namespaces" (formerly known as "internal module") to achieve code isolation.

File: docs/typescript.md [hunk 5]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
+  export function foo() {}
+  function bar() {}  // this function can only be addressed from within the `Module` namespace
+}
+```
+
+```ts
+// anyFiles.ts
+Module.foo();  // address a namespace's exported content directly
+
+const nameIWantForMyImports = Package.foo;  // to simulate `import` with renaming
```
Comment (grant) [inline docs/typescript.md]: Where is `Package` defined?

File: docs/typescript.md [hunk 6]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
-Currently, `clasp` supports [`typescript@2.9.2`](https://www.npmjs.com/package/typescript/v/2.9.2). If there is a feature in a newer  TypeScript version that you'd like to support, or some experimental flag you'd like enabled, please file a bug.
+Currently, `clasp` supports [`typescript@3.3.3`](https://www.npmjs.com/package/typescript/v/3.3.3). If there is a feature in a newer TypeScript version that you'd like to support, or some experimental flag you'd like enabled, please open an issue.
+
+### Modules, exports and imports
+
+Currently, Google Apps Script does not support modules, hence the typical `export`/`import` patern cannot be used and the example below will fail:
+
+```ts
+// module.ts
+export const foo = 'Hello from Module';  // `foo` is added to the `exports` object in the global namespace
```
Comment (grant) [inline docs/typescript.md]: Preview doesn't look nice. Do this:

```ts
// `foo` is added to the `exports` object in the global namespace
export const foo = 'Hello from Module';
```

File: docs/typescript.md [hunk 7]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
+```
+
+```ts
+// main.ts
```
Comment (grant) [inline docs/typescript.md]: - Remove unnecessary whitespace
- Comments should be 1 space after code.

```ts
// main.ts
import { foo } from './module'; // this statement is ignored
const bar = foo; // the variable `foo` does not exist in the global namespace
```

File: docs/typescript.md [hunk 8]
```diff
@@ -201,15 +205,70 @@ Advanced Service currently do not have TypeScript autocompletion.
...
+```
+
+#### The `namespace` statement workaround
+
+Here we take advantage of TypeScript "namespaces" (formerly known as "internal module") to achieve code isolation.
+
+Namespace definition can be nested, spread over multiple files and do not need any `import`/`require` statement to be used.
+
+```ts
+// anyFiles.ts
```
Comment (grant) [inline docs/typescript.md]: `// module.ts`

File: docs/typescript.md (hunk 5 above)
Comment (PopGoesTheWza) [reply docs/typescript.md]: typo... :(

**Comment by PopGoesTheWza** [comment]
@grant for real case use of the `namespace` workaround, you may check at my swgoh repositories

**Comment by grant** [comment]
There are some small mistakes left, but it's fine for now.
Thanks!
